lib/git-lfs-s3/application.rb

```diff
@@ -47,29 +49,174 @@ def protected!
+          },
+        },
+      }
+    end
+
+    def obj_upload(authenticated, obj, obj_json)
+      # Format a single upload object.
+      oid = obj_json[:oid]
+      size = obj_json[:size]
+      {
+        'oid'           => oid,
+        'size'          => size,
+        'authenticated' => authenticated,
+        'actions'       => {
+          'upload'        => {
+            'href'          => obj.presigned_url(:put,
+                                                 acl: 'public-read',
+                                                 :expires_in => 86400),
+            'expires_at'    => expire_at,
+          },
+        },
+      }
+    end
+
+    def obj_error(error, message, obj_json)
+      # Format a single error object.
+      {
+        'oid'         => obj_json[:oid],
+        'size'        => obj_json[:size],
+        'error'       => {
+          'code'        => error,
+          'message'     => message,
+        },
+      }
+    end
+
+    def download(authenticated, params)
+      # Handle git-lfs batch downloads.
+      objects = []
+      params[:objects].each do |obj_json|
+        obj_json = indifferent_params(obj_json)
+        obj = object_data(obj_json[:oid])
+        if valid_obj?(obj_json)
+          if obj.exists?
+            objects.push(obj_download(authenticated, obj, obj_json))
+          else
+            objects.push(obj_error(404, 'Object does not exist', obj_json))
+          end
+        else
+          objects.push(obj_error(422, 'Validation error', obj_json))
+        end
+      end
+      objects
+    end
+
+    def upload(authenticated, params)
+      # Handle git-lfs batch uploads.
+      objects = []
+      params[:objects].each do |obj_json|
+        obj_json = indifferent_params(obj_json)
+        obj = object_data(obj_json[:oid])
+        if valid_obj?(obj_json)
+          if obj.exists?
+            objects.push(obj_download(authenticated, obj, obj_json))
+          else
+            objects.push(obj_upload(authenticated, obj, obj_json))
+          end
+        else
+          objects.push(obj_error(422, 'Validation error', obj_json))
+        end
+      end
+      objects
+      end
+
+    def lfs_resp(objects)
+      # Successful git-lfs batch response.
+      status(200)
+      resp = {
+        'transfer' => 'basic',
+        'objects' => objects
+      }
+      body MultiJson.dump(resp)
+    end
+    
+    def error_resp(status_code, message)
+      # Error git-lfs batch response.
+      status(status_code)
+      resp = {
+        'message' => message,
+        'request_id' => SecureRandom::uuid
+      }
+      body MultiJson.dump(resp)
+    end
+    
+    post '/objects/batch', provides: 'application/vnd.git-lfs+json' do
+      # git-lfs batch API
+      authenticated = authorized?
+      params = indifferent_params(JSON.parse(request.body.read))
+      logger.debug params
+      if params[:operation] == 'download'
+        if settings.public_server
```

Comment:
That should be changed. It should be a 401. Let me test with a GitHub account that doesn't have access.